### DISCUSSION TITLE
Support for other units in paperHeight/Width & margins

### DISCUSSION THREAD

**Discussion by FrankHeijden** [gotenberg/gotenberg#838] [Q&A] [answered]
Would be great if other units such as mm, px, cm could be used! Puppeteer [does support](https://pptr.dev/api/puppeteer.pdfoptions) supplying a string (value + unit), so it would be great if that could be used as well in gotenberg.

**Comment by gulien** [comment]
Hello @FrankHeijden,

See https://github.com/gotenberg/gotenberg/issues/591#issuecomment-1961971483 for the reason of non implementation (for now).

**Reply by FrankHeijden** [reply]
Thanks for the quick reply! Maybe my understanding of that piece of code is wrong, but would defaulting that piece of code to `${value}in` work?

**Reply by gulien** [reply]
Actually, now that I re-see my comment, I think it is doable if I convert other units to inches before entering any Chromium related tasks.

**Reply by gulien** [reply]
I’ve reopened the issue, no ETA though 😅

**Reply by FrankHeijden** [reply]
For now I will just convert my values to inches like so:
```typescript
export interface UnitValue {
  value: number;
  unit: 'pt' | 'px' | 'in' | 'mm' | 'cm' | 'pc';
}

export const unitConversions: { [U in UnitValue['unit']]: number } = {
  pt: 1 / 72,
  px: 1 / 96,
  in: 1,
  mm: 1 / 25.4,
  cm: 1 / 2.54,
  pc: 1 / 6,
};

export function parseUnitValueAsInches(uv: UnitValue): number {
  return uv.value * unitConversions[uv.unit];
}
```

The only issue is that for example with the relative units like em, ex, I cannot calculate them directly.

**Reply by gulien** [reply]
Thanks for the details! I think I'll avoid relative units, too many caveats.

**Reply by FrankHeijden** [reply]
Hm @gulien in the commit [eb05414](https://github.com/gotenberg/gotenberg/commit/eb054146d27aadbe8824e2f11c39bfaf030ecdbd), I see that you also use the above conversion logic for units; is there a hard requirement in Gotenberg to know the exact measurements before using it as input to Puppeteer? The current implementation would still not work for dynamic/relative CSS units like `em` and `ex`.

**Reply by gulien** [reply, marked as answer]
@FrankHeijden `gotenberg/gotenberg:edge` (https://github.com/gotenberg/gotenberg/actions/runs/8315842453/job/22754688797) should soon be available with support for this feature. For instance:

* `15pt`
* `8.25in` or `8.25`
* `300px`
* etc.

**Reply by gulien** [reply]
Yep as I said, I don't want to deal with relative units. What use cases do you have in mind?

**Reply by gulien** [reply]
> Gotenberg to know the exact measurements before using it as input to Puppeteer

I'm not using Puppeteer 😅

**Reply by FrankHeijden** [reply]
We're using pdf generation for barcodes; I could see dynamic sizes related to the font size being useful there if clients don't use a fixed size paper width/height. What's the reason that Gotenberg needs to calculate the exact sizes of the paper/width before entering it into puppeteer?

**Reply by gulien** [reply]
Again, I'm not using Puppeteer, but the Chrome DevTools protocol. No idea how Puppeteer works with relative units 🤔

**Reply by gulien** [reply]
A workaround may be to try the `preferCssPageSize` form field and put the page size in CSS.

**Reply by FrankHeijden** [reply]
Oh I am so sorry; I think I might have misread the [documentation](https://gotenberg.dev/docs/troubleshooting#chromium). I just read this piece again:
> Gotenberg internally employs the same **protocol** as [Puppeteer](https://pptr.dev/) for communication with Chromium.

The key word being "protocol" 😅

**Reply by FrankHeijden** [reply]
I looked into the puppeteer documentation again, and they do not seem to mention which units they support either. The [PDFOptions](https://pptr.dev/api/puppeteer.pdfoptions#properties) they define just mentions "You can pass in a number or a string with a unit.". 

Upon taking a further look at the actual code, I also just found that they [do a simple conversion to px, and then to inches](https://github.com/puppeteer/puppeteer/blob/9aef7dc5e5d42086a08f9f27fa1467ec41219201/packages/puppeteer-core/src/common/util.ts#L404-L446); as it seems they also just use the Chrome Devtools protocol's [Page](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-printToPDF). It seems they default to a value of 0 inches when a unit is not recognized.

So I guess that also resolves the question how puppeteer does it, they don't 😅
